src/Analyzer/InDepthQueryTreeVisitor.h

```diff
@@ -84,4 +91,144 @@ class InDepthQueryTreeVisitor
 template <typename Derived>
 using ConstInDepthQueryTreeVisitor = InDepthQueryTreeVisitor<Derived, true /*const_visitor*/>;
 
+/** Same as InDepthQueryTreeVisitor and additionally keeps track of current scope context.
+  * This can be useful if your visitor has special logic that depends on current scope context.
+  */
+template <typename Derived, bool const_visitor = false>
+class InDepthQueryTreeVisitorWithContext
+{
+public:
+    using VisitQueryTreeNodeType = std::conditional_t<const_visitor, const QueryTreeNodePtr, QueryTreeNodePtr>;
+
+    explicit InDepthQueryTreeVisitorWithContext(ContextPtr context)
+        : current_context(std::move(context))
+    {}
+
+    /// Return true if visitor should traverse tree top to bottom, false otherwise
+    bool shouldTraverseTopToBottom() const
+    {
+        return true;
+    }
+
+    /// Return true if visitor should visit child, false otherwise
+    bool needChildVisit(VisitQueryTreeNodeType & parent [[maybe_unused]], VisitQueryTreeNodeType & child [[maybe_unused]])
+    {
+        return true;
+    }
+
+    const ContextPtr & getContext() const
+    {
+        return current_context;
+    }
+
+    const ContextPtr & getContext()
+    {
+        return current_context;
+    }
+
+    const Settings & getSettings() const
+    {
+        return current_context->getSettingsRef();
+    }
+
+    const Settings & getSettings()
+    {
+        return current_context->getSettingsRef();
+    }
+
+    void visit(VisitQueryTreeNodeType & query_tree_node)
+    {
+        auto current_scope_context_ptr = current_context;
+        SCOPE_EXIT(
+            current_context = std::move(current_scope_context_ptr);
+        );
+
+        if (auto * query_node = query_tree_node->template as<QueryNode>())
+            current_context = query_node->getContext();
+        else if (auto * union_node = query_tree_node->template as<UnionNode>())
+            current_context = union_node->getContext();
+
+        bool traverse_top_to_bottom = getDerived().shouldTraverseTopToBottom();
+        if (!traverse_top_to_bottom)
+            visitChildren(query_tree_node);
+
+        getDerived().visitImpl(query_tree_node);
+
+        if (traverse_top_to_bottom)
+            visitChildren(query_tree_node);
+    }
+private:
+    Derived & getDerived()
```

Comment:
Same here

Reply:
This is different, because it will return mutable version of subclass. It is important, because subclass visitImpl can modify its state.